app/types/app-extension.d.ts

```diff
@@ -0,0 +1,61 @@
+import { IResolve } from './app-paths'
+
+export interface IndexAPI {
+  ctx: Record<string, unknown>
+  extId: string
+  prompts: Record<string, unknown>
+  resolve: IResolve,
+  appDir: string,
+  // __hooks: IndexAPIHooks,
+  getPersistentConf: () => Record<string, unknown>,
+  setPersistentConf: (cfg: Record<string, unknown>) => void,
+  mergePersistentConf: (cfg: Record<string, unknown>) => void,
+  compatibleWith: (packageName: string, semverCondition?: string) => void,
+  hasPackage: (packageName: string, semverCondition?: string) => boolean,
+  hasExtension: (extId: string) => boolean,
+  getPackageVersion: (packageName: string) => string|undefined
+  extendQuasarConf: (fn: Function) => void,
+  chainWebpack: (fn: Function) => void,
```

Comment:
You should be able to type these using `QuasarConf`, as these are proxies to those methods
Here's an helper you could use to extract parameters of those functions and how to use it

Eg.
```ts
// `DeepRequired` and `DeepNonNullable` are taken from `ts-essentials`, which you can add as a dev dependency to q/app

type QuasarConfProxy = DeepRequired<DeepNonNullable<QuasarConf>>;
type ExtractQuasarConfParameters<
  FirstLevelKey extends keyof QuasarConfProxy,
  SecondLevelKey extends keyof QuasarConfProxy[FirstLevelKey],
  MaybeFunction = QuasarConfProxy[FirstLevelKey][SecondLevelKey]
> = MaybeFunction extends (...args: any) => any
  ? Parameters<MaybeFunction>
  : never;

chainWebpack: (fn: (...args: [...ExtractQuasarConfParameters<"build", "chainWebpack">, IndexAPI]) => void) => void,
extendWebpack: (fn: (...args: [...ExtractQuasarConfParameters<"build", "extendWebpack">, IndexAPI]) => void) => void,
```

It seems like the second parameter is always the `isClient/isServer` object, even for webpack methods which shouldn't care about it. I guess this a technical limitation. It can be added manually or we can expose `InvokeParams` from q/app directly

Reply:
According to the [docs](https://quasar.dev/app-extensions/development-guide/index-api) it is the same for all methods, however for the SSR methods the added comment is ` // isClient is always "false" and isServer is always "true"`

Reply:
It is the same into AE Index API, but it isn't into the `QuasarConf` type, that's where things get tricky and we may need to tamper with the returned type extracted from `QuasarConf`, I don't think all have the same parameters (but I checked this nearly 3 weeks ago, I may be wrong)

Reply:
The extracted types are
```
type extendWebpack = (fn: (args_0: WebpackConfiguration, args_1: InvokeParams, args_2: IndexAPI) => void) => void
type chainWebpack = (fn: (args_0: Config, args_1: InvokeParams, args_2: IndexAPI) => void) => void
```
which seems correct? QuasarConf['build'] extendWebpack and chainWebpack are also statically typed: https://github.com/quasarframework/quasar/blob/87f7cd257e361e9bb79d189cc862018585361bd4/app/types/configuration/build.d.ts#L38-L46

Reply:
You're right, I got confused by the fact that Mode specific `extend/chainWebpack` methods don't have `InvokeParams`, but since you're using the default ones for all signatures, it should be good, thanks for following up